lasagne/layers/normalization.py

```diff
@@ -373,3 +375,158 @@ def batch_norm(layer, **kwargs):
         nonlin_name = bn_name and bn_name + '_nonlin'
         layer = NonlinearityLayer(layer, nonlinearity, name=nonlin_name)
     return layer
+
+
+class InstanceNormLayer(Layer):
+    """
+    Instance Normalization
+
+    This layer implements instance normalization of its inputs:
+
+    .. math::
+        y_i = \\frac{x_i - \\mu_i}{\\sqrt{\\sigma_i^2 + \\epsilon}}
+
+    That is, each input example (instance) is normalized to zero mean
+    and unit variance. In contrast to batch normalization, the mean and
+    variance is usually taken per example, and not across examples,
+    so the same operation can be applied during training and testing.
+    During both training and testing, :math:`\\mu_i` and
+    :math:`\\sigma_i^2` are defined to be the mean and variance
+    of the instances :math:`i` in the current input mini-batch :math:`x`.
+    The advantages of using this implementation over e.g.
+    :class:`BatchNormLayer` with adapted axes arguments, are its
+    independence of the input size, as no parameters are learned and stored.
+
+    Parameters
+    ----------
+    incoming : a :class:`Layer` instance or a tuple
+        The layer feeding into this layer, or the expected input shape
+    axes : 'auto', int or tuple of int
+        The axis or axes to normalize over. If ``'auto'`` (the default),
+        normalize over all axes except for the first two:
+        this will normalize over all spatial dimensions for
+        convolutional layers.
```

Comment:
With the new `StandardizationLayer`, it might make sense to have `'auto'` normalize over the second axis for 2D input, and over all but the first two axes for 3D+ input. This would do something useful for dense layers, recurrent layers and convolutional layers.